sdk/metric/pipeline.go

```diff
@@ -178,96 +157,181 @@ func (p *pipeline) produce(ctx context.Context) (metricdata.ResourceMetrics, err
 
 // inserter facilitates inserting of new instruments into a pipeline.
 type inserter[N int64 | float64] struct {
+	cache    instrumentCache[N]
 	pipeline *pipeline
 }
 
-func newInserter[N int64 | float64](p *pipeline) *inserter[N] {
-	return &inserter[N]{p}
+func newInserter[N int64 | float64](p *pipeline, c instrumentCache[N]) *inserter[N] {
+	return &inserter[N]{cache: c, pipeline: p}
 }
 
-// Instrument inserts instrument inst with instUnit returning the Aggregators
-// that need to be updated with measurments for that instrument.
+// Instrument inserts the instrument inst with instUnit into a pipeline. All
+// views the pipeline contains are matched against, and any matching view that
+// creates a unique Aggregator will be inserted into the pipeline and included
+// in the returned slice.
+//
+// The returned Aggregators are ensured to be deduplicated and unique. If
+// another view in another pipeline that is cached by this inserter's cache has
+// already inserted the same Aggregator for the same instrument, that
+// Aggregator instance is returned.
+//
+// If another instrument has already been inserted by this inserter, or any
+// other using the same cache, and it conflicts with the instrument being
+// inserted in this call, an Aggregator matching the arguments will still be
+// returned but an Info level log message will also be logged to the OTel
+// global logger.
+//
+// If the passed instrument would result in an incompatible Aggregator, an
+// error is returned and that Aggregator is not inserted or returned.
+//
+// If an instrument is determined to use a Drop aggregation, that instrument is
+// not inserted nor returned.
 func (i *inserter[N]) Instrument(inst view.Instrument, instUnit unit.Unit) ([]internal.Aggregator[N], error) {
-	seen := map[instrumentID]struct{}{}
 	var aggs []internal.Aggregator[N]
 	errs := &multierror{wrapped: errCreatingAggregators}
+	// The cache will return the same Aggregator instance. Use this fact to
+	// compare pointer addresses to deduplicate Aggregators.
+	seen := make(map[internal.Aggregator[N]]struct{})
```

Comment:
This means we are relying on the (somewhat) unstated idea that `internal.Aggregator`s are going to be references.  I'm ok with that assumption, as I think it will be almost universal, but we should probably document that with the interface for those who come after.

Reply:
Given this is a type from an internal package it seemed like something we could indeed guarantee.

Where were you thinking of adding documentation for this?

Reply:
@MadVikingGod I added https://github.com/open-telemetry/opentelemetry-go/pull/3251/commits/1f836588b9d8ba37638e255f270b86085b485aed, does that address this?

Reply:
I was thinking more a comment around https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/metric/internal/aggregator.go#L28-L29. But tests don't hurt.

Reply:
Updated.